src/datasets/arrow_dataset.py

```diff
@@ -2145,6 +2145,55 @@ def rename(columns):
         dataset._fingerprint = new_fingerprint
         return dataset
 
+    @transmit_tasks
+    @transmit_format
+    @fingerprint_transform(inplace=False)
+    def select_columns(self, column_names: Union[str, List[str]], new_fingerprint: Optional[str] = None) -> "Dataset":
+        """Select one or several column(s) in the dataset and the features
+        associated to them.
+
+        Args:
+            column_names (`Union[str, List[str]]`):
+                Name of the column(s) to keep.
+            new_fingerprint (`str`, *optional*):
+                The new fingerprint of the dataset after transform. If `None`,
+                the new fingerprint is computed using a hash of the previous
+                fingerprint, and the transform arguments.
+
+        Returns:
+            [`Dataset`]: A copy of the dataset object which only consists of
+            selected columns.
+
+        Example:
+
+        ```py
+        >>> from datasets import load_dataset
+        >>> ds = load_dataset("rotten_tomatoes", split="validation")
+        >>> ds.select_columns('text')
+        Dataset({
+            features: ['text'],
+            num_rows: 1066
+        })
+        ```
+        """
+        if isinstance(column_names, str):
+            column_names = [column_names]
+        for column_name in column_names:
+            if column_name not in self._data.column_names:
+                raise ValueError(
+                    f"Column name {column_name} not in the "
+                    "dataset. Current columns in the dataset: "
+                    f"{self._data.column_names}."
+                )
+
+        dataset_columns = frozenset(column_names)
+        dataset = copy.deepcopy(self)
+        dataset._info.features = {k: v for k, v in dataset._info.features.items() if k in dataset_columns}
+        dataset._data = dataset._data.select(column_names)
+        dataset._data = update_metadata_with_features(dataset._data, dataset.features)
+        dataset._fingerprint = new_fingerprint
+        return dataset
```

Comment:
Some nits to be consistent with `remove_columns`:

```suggestion
        dataset = copy.deepcopy(self)
        if isinstance(column_names, str):
            column_names = [column_names]

        for column_name in column_names:
            if column_name not in self._data.column_names:
                raise ValueError(
                    f"Column name {column_name} not in the "
                    "dataset. Current columns in the dataset: "
                    f"{self._data.column_names}."
                )

        for column_name in column_names:
            del dataset._info.features[column_name]

        dataset._data = dataset._data.select(column_names)
        dataset._data = update_metadata_with_features(dataset._data, dataset.features)
        dataset._fingerprint = new_fingerprint
        return dataset
```

Reply:
> totally unclear how memory is managed

It's always unclear how memory is managed since Python relies on a garbage collector.

I'm OK with having `deepcopy` after argument validation, but let's drop `frozenset` as it seems unnecessary for the simple dict comprehension.

Reply:
>  but let's drop frozenset

Sure. I don't insist on using sets, however we should keep in mind that this will result in poor scalability in theory and on practice. I've made a mini benchmark in case we will return to this in the future
```python
In [1]: from random import sample
   ...: from string import ascii_letters
   ...: 
   ...: # Mean length of english word is about 4-5 chars.
   ...: keys = [''.join(sample(ascii_letters, 5)) for _ in range(1000)]
   ...: 
   ...: subset_lst = keys[:4]
   ...: subset_set = set(subset_lst)
   ...: subset_fro = frozenset(subset_lst)  # The same as set but it shows intent.
   ...: 
   ...: # Small number of columns.
   ...: %timeit {k: None for k in keys[:10] if k in subset_lst}
   ...: %timeit {k: None for k in keys[:10] if k in subset_set}
   ...: %timeit {k: None for k in keys[:10] if k in subset_fro}
   ...: 
   ...: # Large number of columns (wide table).
   ...: %timeit {k: None for k in keys if k in subset_lst}
   ...: %timeit {k: None for k in keys if k in subset_set}
   ...: %timeit {k: None for k in keys if k in subset_fro}
   ...: 
836 ns ± 5.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
541 ns ± 2.91 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
561 ns ± 2.72 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
80.2 µs ± 14.4 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
48.1 µs ± 2.83 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
43.6 µs ± 416 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
This benchmark is a bit off since it doesn't include the time needed to instantiate a `set`/`frozenset`.

Also, these dataset functions are not meant to be run in the for loop, so a slight overhead is OK (plus, in Python unless it's an order of magnitude slower, readability >> performance)

Reply:
Right. It should demonstrate a typical timing. Construction time of a small `set` is a order of magnitude of list construction. Sets behave like lists for a few columns but scales much better.
```python
In [2]: %timeit list(subset_lst)
   ...: %timeit set(subset_lst)
   ...: %timeit frozenset(subset_lst)
66.4 ns ± 1.04 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
108 ns ± 3.32 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
109 ns ± 2.07 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
And "readability" argument is always about habits and a matter of taste.